onnxruntime/core/providers/coreml/coreml_execution_provider.cc

```diff
@@ -57,7 +58,32 @@ CoreMLExecutionProvider::GetCapability(const onnxruntime::GraphViewer& graph_vie
       [&]() {
         HashValue model_hash;
         int metadef_id = metadef_id_generator_.GenerateId(graph_viewer, model_hash);
-        return MakeString(COREML, "_", model_hash, "_", metadef_id);
+        std::string user_provided_key;
+        const Graph* main_graph = &graph_viewer.GetGraph();
+        while (main_graph->IsSubgraph()) {
+          main_graph = main_graph->ParentGraph();
+        }
+        if (main_graph->GetModel().MetaData().count(kCOREML_CACHE_KEY) > 0) {
+          user_provided_key = graph_viewer.GetGraph().GetModel().MetaData().at(kCOREML_CACHE_KEY);
+          if (user_provided_key.size() > 64 ||
+              std::any_of(user_provided_key.begin(), user_provided_key.end(),
+                          [](unsigned char c) { return !std::isalnum(c); })) {
+            LOGS(logger, ERROR) << "[" << kCOREML_CACHE_KEY << ":" << user_provided_key << "] is not a valid cache key."
+                                << " It should be alphanumeric and less than 64 characters.";
+          }
+          // invalid cache-key
+          if (user_provided_key.size() == 0) {
```

Comment:
Can we do this once outside of the gen_metadef_name lambda?

When there's an error we're logging it, but nothing is setting user_provided_key to empty, so not clear how the 'invalid cache-key' `if` condition will be satisfied.

Reply:
Sure， Sounds good.